src/lib/Imp.hs

```diff
@@ -127,15 +148,30 @@ translateDecl env (maybeDest, (Let _ b bound)) = do
 translateExpr :: SubstEnv -> WithDest Expr -> ImpM Atom
 translateExpr env (maybeDest, expr) = case expr of
   Hof hof     -> toImpHof env (maybeDest, hof)
-  App x' idx' -> case getType x' of
-    TabTy _ _ -> do
-      x <- impSubst env x'
-      idx <- impSubst env idx'
-      case x of
-        Lam a@(Abs _ (TabArrow, _)) ->
-          translateBlock mempty (maybeDest, snd $ applyAbs a idx)
-        _ -> error $ "Invalid Imp atom: " ++ pprint x
-    _ -> error $ "shouldn't have non-table app left"
+  App f' x' -> do
+    f <- impSubst env f'
+    x <- impSubst env x'
+    case getType f of
+      TabTy _ _ -> do
+        case f of
+          Lam a@(Abs _ (TabArrow, _)) ->
+            translateBlock mempty (maybeDest, snd $ applyAbs a x)
+          _ -> error $ "Invalid Imp atom: " ++ pprint f
+      _ -> do
```

Comment:
Can you add a comment, saying that this is for functions that we haven't inlined?